src/collapse/mod.rs

```diff
@@ -69,3 +85,183 @@ pub trait Collapse {
+            let map = CHashMap::with_capacity(CAPACITY_MULTI_THREADED_HASHMAP);
+            let arc = Arc::new(map);
+            Occurrences::MultiThreaded(arc)
+        }
+    }
+
+    fn add(&mut self, key: String, count: usize) {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => *map.entry(key).or_insert(0) += count,
+            MultiThreaded(arc) => {
+                arc.upsert(key, || count, |v| *v += count);
+            }
+        }
+    }
+
+    fn is_concurrent(&self) -> bool {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(_) => false,
+            MultiThreaded(_) => true,
+        }
+    }
+
+    fn write<W>(&mut self, mut writer: W) -> io::Result<()>
+    where
+        W: io::Write,
+    {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(ref mut map) => {
+                let mut contents: Vec<_> = map.drain().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+            MultiThreaded(ref mut arc) => {
+                let map = match Arc::get_mut(arc) {
+                    Some(map) => map,
+                    None => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::Other,
+                            "Attempting to drain contents of a concurrent \
+                             hashmap when more than one thread has access to it, \
+                             which is not allowed.",
+                        ))
+                    }
+                };
+                let map = mem::replace(
+                    map,
+                    CHashMap::with_capacity(CAPACITY_MULTI_THREADED_HASHMAP),
+                );
+                let mut contents: Vec<_> = map.into_iter().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+struct Chunks<'a> {
+    current_index: usize,
+    data: &'a [u8],
+    indices: &'a [usize],
+}
+
+impl<'a> Iterator for Chunks<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let chunk = {
+            let start = self.indices.get(self.current_index)?;
+            match self.indices.get(self.current_index + 1) {
+                Some(end) => &self.data[*start..*end],
+                None => &self.data[*start..],
+            }
+        };
+        self.current_index += 1;
+        Some(chunk)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/// Representation of input data that can be chunked up and shared
+/// across threads easily.
+struct Input {
```

Comment:
I'd love some better docs on `Input`, since someone implementing a new `Collapser` will have to understand it.

Reply:
Good idea. That is, if `Input` remains after addressing the comment about sending chunks in a streaming fashion as opposed to going over the entire file first (which is a good idea and worth testing).